crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -0,0 +1,878 @@
+                }
+
+                continue;
+            }
+
+            // Check if the checkpoint is the one we are expecting next!
+            // if next_checkpoint == checkpoint.checkpoint.sequence_number {
+            // Try to upgrade the signed checkpoint to a certified one if possible
+            if state_checkpoints
+                .lock()
+                .handle_checkpoint_certificate(&checkpoint, &None)
+                .is_err()
+            {
+                // One of the errors may be due to the fact that we do not have
+                // the full contents of the checkpoint. So we try to download it.
+                // TODO: clean up the errors to get here only when the error is
+                //       "No checkpoint set at this sequence."
+                if let Ok(contents) = get_checkpoint_contents(
+                    active_authority.state.name,
+                    active_authority.net.clone(),
+                    &checkpoint,
+                )
+                .await
+                {
+                    // Retry with contents
+                    let _ = state_checkpoints
+                        .lock()
+                        .handle_checkpoint_certificate(&checkpoint, &Some(contents));
+                }
+            }
+        }
+
+        // (3) Process any unprocessed transactions. We do this before trying to move to the
+        //     next proposal.
+        if let Err(err) = process_unprocessed_digests(
+            active_authority,
+            state_checkpoints.clone(),
+            timing.per_other_authority_delay,
+        )
+        .await
+        {
+            warn!("Error processing unprocessed: {:?}", err);
+            // Nothing happens until we catch up with the unprocessed transactions of the
+            // previous checkpoint.
+            continue;
+        }
+
+        // (4) Check if we need to advance to the next checkpoint, in case >2/3
+        // have a proposal out. If so we start creating and injecting fragments
+        // into the consensus protocol to make the new checkpoint.
+        let weight: usize = proposals
+            .iter()
+            .map(|(auth, _)| active_authority.state.committee.weight(auth))
+            .sum();
+
+        let _start_checkpoint_making = weight > active_authority.state.committee.quorum_threshold();
+
+        let proposal = state_checkpoints.lock().new_proposal().clone();
+        if let Ok(my_proposal) = proposal {
+            diff_proposals(
+                active_authority,
+                state_checkpoints.clone(),
+                &my_proposal,
+                proposals,
+                timing.consensus_delay_estimate,
+            )
+            .await;
+        }
+
+        // (5) Wait for a long long time.
+        let name = state_checkpoints.lock().name;
+        let next_checkpoint = state_checkpoints.lock().next_checkpoint();
+
+        debug!("{:?} at checkpoint {:?}", name, next_checkpoint);
```

Comment:
This may remind you of Python3 :)
```suggestion
        debug!("{name:?} at checkpoint {next_checkpoint:?}");
```